### PR TITLE
CNS-document-top-architecture

### PR DESCRIPTION
**Pull request by omerlavanet**

### CODE REVIEW

**Comment by nimrod-teich** [comment]
:raised_hands:

File: x/README.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+When staking a provider, it's mandatory to simultaneously select a validator for delegation. Validator delegation follows the regular cosmos-sdk delegation process, where delegators participate in validator rewards post-commission. Provider delegations also contribute to rewards and boost the effective stake of the provider, attracting more consumers and increasing both load and rewards. These provider delegations are influenced by a provider commission and a delegation limit. In the absence of a chosen validator, our CLI automatically selects the largest validator or a validator with existing delegations.
+
+### Adding Specifications
+
+In Lava, governance has the authority to vote on the addition of a new API specification through a spec add proposal. A specification, represented by a detailed configuration JSON, defines a service comprehensively.
```
Comment (oren-lava) [inline x/README.md:94]: for clarity: spec add -> "spec-add"

File: x/README.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+When delegating to a validator, it's optional to select a provider. If no provider is chosen, the funds remain available for future delegation to a provider in a designated "empty_provider" bucket. While this might not maximize validators' staking effects and rewards, it can be remedied with a delegation transaction. Lava supports cosmos basic staking messages, and through hooks, dual staking entries are adapted accordingly. Validator self-staking is considered self-delegation and can benefit from provider delegation.
+
+When staking a provider, it's mandatory to simultaneously select a validator for delegation. Validator delegation follows the regular cosmos-sdk delegation process, where delegators participate in validator rewards post-commission. Provider delegations also contribute to rewards and boost the effective stake of the provider, attracting more consumers and increasing both load and rewards. These provider delegations are influenced by a provider commission and a delegation limit. In the absence of a chosen validator, our CLI automatically selects the largest validator or a validator with existing delegations.
+
+### Adding Specifications
```
Comment (oren-lava) [inline x/README.md:92]: you mentioned it above, but maybe mention here again the contributor role?

File: x/README.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+There exist three tiers of policies, where the strictest policy takes precedence. While a subscription cannot surpass the constraints outlined by the plan, it can lower certain limits, such as the provider count, within the given policy framework.
+
+### Pairing
+
+Consumers possessing a valid subscription or project can engage with providers to access APIs. However, full communication across the entire Lava network is restricted. Lava employs a pseudo-random matching algorithm between consumers and providers, considering factors such as providers' stake, geolocation, Excellence Quality of Service metric, supported addons, and more, along with specific consumer parameters. This algorithm pairs them within a particular epoch. Every epoch, this selection undergoes changes to maintain a fair and balanced distribution, ensuring privacy by diversifying traffic and preventing censorship or spear phishing attempts.
```
Comment (oren-lava) [inline x/README.md:108]: you didn't mention projects until now. Maybe add a small clarification note somewhere?

File: x/README.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+Only providers within a consumer's designated pairing can claim rewards. The maximum amount of Compute Units (CU) they can claim within a specific plan decreases with an increasing number of providers. This design ensures a fair distribution of the workload among providers, preventing an overwhelming concentration of requests on a single provider.
+
+The pairing mechanism is not computed on-chain until rewards are claimed. Consumers execute it off-chain to view available providers, while providers run it to validate a consumer's request for the first time. The chain executes it upon aggregated payment requests, ensuring accurate validation and reward distribution based on the established pairing.
+
+### Lava Protocol P2P
+
+lava defines a protocol between consumers and providers.
```
Comment (oren-lava) [inline x/README.md:116]: lava -> Lava

File: x/README.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+Providers locally store the latest relay request, containing a cryptographic signature from the consumer developer or subscription key, as proof. When an epoch concludes, after a delay, these rewards aggregate into a transaction termed 'relay_payment,' automatically triggered by the provider service. This transaction passes verifications and accumulates compute units (CU) on behalf of the provider.
+
+Upon a subscription's termination, all provided CU (capped as per the plan policy, both total and per epoch) are factored into dividing rewards among providers. Each provider receives their portion from the subscription's cost after deducting validators' and contributors' participation. Subsequently, these rewards are divided post-commission with delegators. Consequently, provider rewards are distributed monthly per subscription, contingent upon its purchase date. Rewards for providers and their delegators can be claimed through a dualstaking transaction.
```
Comment (oren-lava) [inline x/README.md:142]: not termination, but end of month.

Also, the paragraph is not clear on distributing rewards until the last sentence. It looks like the providers immediately get rewarded (by the way you put it) and only the last sentence clarifies that provider rewards are basically delegation rewards. Maybe change the phrasing a bit to emphasize that the rewards are allocated on chain and need to be claimed with a specific TX

File: x/README.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+### Conflicts
+
+When utilizing the services, consumers leverage the protocol's data reliability feature to cross-verify deterministic responses for identical queries from multiple providers. In cases of discrepancies, cryptographically signed by both providers, consumers can initiate a conflict transaction to report the fraudulent provider. This action triggers an on-chain resolution process involving a commit-and-reveal vote by all selected providers within the spec (known as the jury mechanism). Providers selected for this process are obligated to participate; failure to do so results in potential penalties. Following the necessary votes, the provider found at fault will face repercussions, potentially leading to their temporary restriction. As part of our ongoing testing and refinement process, we aim to enhance this mechanism, particularly for specs with substantial stake backing for majority trust.
```
Comment (oren-lava) [inline x/README.md:156]: can the providers know in advance which query will be used to test their answer? what prevents from a provider to return valid responses to these particular queries and lie in other ones?

File: x/README.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+## Modules
+
+### utility
```
Comment (oren-lava) [inline x/README.md:160]: utility -> Utility

File: x/README.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+Only providers within a consumer's designated pairing can claim rewards. The maximum amount of Compute Units (CU) they can claim within a specific plan decreases with an increasing number of providers. This design ensures a fair distribution of the workload among providers, preventing an overwhelming concentration of requests on a single provider.
+
+The pairing mechanism is not computed on-chain until rewards are claimed. Consumers execute it off-chain to view available providers, while providers run it to validate a consumer's request for the first time. The chain executes it upon aggregated payment requests, ensuring accurate validation and reward distribution based on the established pairing.
+
+### Lava Protocol P2P
```
Comment (oren-lava) [inline x/README.md:114]: Shouldn't you mention the constant availability feature? It's an amazing feature that should get more attention

File: x/README.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+#### timerstore
+
+The timerstore module functions as a utility that facilitates block time - based callbacks triggered at the beginning of a block. This functionality enables various processes such as periodic checks, timers for subscription expiry, monthly payouts, and other time-sensitive operations.
+
+#### downtime
+
+The downtime utility module serves to offer insights into the block average time and detects occurrences of blockchain downtime.
+
+### governance
```
Comment (oren-lava) [inline x/README.md:178]: governance -> Governance

File: x/README.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+#### plans
+
+The plans module establishes a framework for governance to regulate available subscription options and their associated policies. Governance has the authority to introduce plans for consumers to purchase at predefined token prices with specific configurations. Additionally, it allows for the modification of existing plans that have already been acquired.
+
+#### protocol
+
+This module is responsible for storing the target and minimum versions for the protocol modules. Any alterations to these versions are possible through a governance parameters proposal.
+
+### consumers & providers
```
Comment (oren-lava) [inline x/README.md:192]: consumers & providers -> Consumers & providers

File: x/README.md [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+This module enables consumers to purchase and engage with their subscriptions. It generates a subscription object for each consumer, managed with monthly timers. Subscriptions retain references for purchased plans, releasing them upon deletion. Upon creation, a default project is initialized, and additional projects can be added via this interface, stored, and managed within the projects module.
+
+The subscription module is also responsible for compensating providers based on accumulated CUs once a subscription expires (or a month change is triggered). Accumulated CUs per provider per chain are stored and utilized when the expiry timer is triggered.
+
+#### projects
+
+This module manages projects intended for developers' usage. It serves as a tool to handle internal usage within user accounts, allowing different allocations of CUs and policy specifications. Each project can define developer and admin keys, along with its specific policy. Subscriptions hold a higher hierarchy and can define policies for their associated projects.
+
+#### pairing module
```
Comment (oren-lava) [inline x/README.md:204]: delete "module" to match other headlines

File: x/README.md [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+This core module serves multiple roles:
+
+* Handles provider staking and their storage, incorporating changes in providers, addition of endpoints, or supported provider specifications within this module. It utilizes epochstorage to snapshot the provider's storage per epoch change.
+* Manages the pairing function between consumers and providers based on policies and available providers.
+* Manages (Quality of Service) QoS reports and payment requests from providers, storing them to prevent double spending and forwarding accumulated CUs to the subscription module.
+
+### tokenomics
```
Comment (oren-lava) [inline x/README.md:212]: tokenomics -> Tokenomics

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/README.md [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+## Roles
+
+### Validator
+
+Validators play a pivotal role in the Lava Blockchain, akin to the classic cosmos-sdk validator role. Their primary function involves producing blocks, validating transactions, and earning rewards from transaction fees and block creation. Beyond this foundational role, validators also actively participate in subscription rewards. As subscriptions increase in number or activity, validator incentives are directly impacted, creating a symbiotic relationship between network growth and validator rewards.
+
+### Consumer
+
+Consumers within the Lava ecosystem are individuals seeking to utilize APIs. They gain access through valid periodic subscriptions acquired either by themselves or by other users using tokens. Active subscriptions empower consumers with a curated list of available providers offering APIs and enable them to allocate rewards. Subscription owners hold the authority to delegate usage within their subscription, facilitating developers' access via designated projects.
```
[the remark this answers is not in the document]
Comment (omerlavanet) [reply x/README.md:58]: this is incorrect, it's not only rpc apis, also other apis

File: x/README.md (hunk 6 above)
Comment (omerlavanet) [reply x/README.md:156]: every query...

Comment (oren-lava) [reply x/README.md]: I read the paragraph again and it still seems to talk about fraud. Can you clarify this mechanism is for downtime? Also, if not here, where do you talk about fraud?

File: x/README.md (hunk 11 above)
Comment (omerlavanet) [reply x/README.md]: it's because there's pairing in concepts and you don't want repetitions

File: x/README.md (hunk 6 above)
Comment (omerlavanet) [reply x/README.md]: vscode moved the comment a bit upwards to reports, i added the fix in conflict

**Review by Yaroms — APPROVED** [review]

**Pull request merged**